### PR TITLE
feat: Adds port/ protocol support for host

### PR DESCRIPTION
**Pull request by walker233**
This update extends the host resource to use correct
port and protocol for the redhat and debian families.
It uses a feature from bash and the timeout command
from the coreutils packages.

When port is specified but protocol is omitted, it
assumes tcp is what was intended. When neither port
nor protcol is specified it falls back to ping the
host.

Tries to address issue #1439 to some degree.

### CODE REVIEW

**Comment by walker233** [comment]
I'm pretty sure that the one failed test job in Travis CI is from the modification. The error and failure should be investigated to that other do no fail in the same way.

File: lib/resources/host.rb [hunk 1]
```diff
@@ -121,12 +121,20 @@ def resolve(hostname)
...
+    def ping(hostname, port = nil, proto = nil)
+      # assume tcp if port is given without corresponding protocol
+      if port && !proto
+        proto = 'tcp'
+      end
+      # use bash builtin capability for checking connectivity
+      # for those os's with timeout from coreutils packages
+      # TODO: Verify whether this works on other linux versions
+      if port && (inspec.os.redhat? || inspec.os.debian?)
+        resp = inspec.command("timeout 1 bash -c 'cat </dev/null >/dev/#{proto}/#{hostname}/#{port}'")
```
Comment (chris-rock) [inline lib/resources/host.rb:129]: Should we check if bash is available? eg. via `command('bash').exist?`

File: lib/resources/host.rb [hunk 2]
```diff
@@ -121,12 +121,20 @@ def resolve(hostname)
...
   class LinuxHostProvider < HostProvider
-    # ping is difficult to achieve, since we are not sure
-    def ping(hostname, _port = nil, _proto = nil)
-      # fall back to ping, but we can only test ICMP packages with ping
-      # therefore we have to skip the test, if we do not have everything on the node to run the test
-      ping = inspec.command("ping -w 1 -c 1 #{hostname}")
-      ping.exit_status.to_i != 0 ? false : true
+    def ping(hostname, port = nil, proto = nil)
+      # assume tcp if port is given without corresponding protocol
+      if port && !proto
```
Comment (chris-rock) [inline lib/resources/host.rb]: should we just set the default to `tcp`?
```
def ping(hostname, port = nil, proto = 'tcp')
```

**Review by chris-rock — COMMENTED** [review]
Thank you @walker233 for this improvement!

File: lib/resources/host.rb (hunk 2 above)
Comment (walker233) [reply lib/resources/host.rb]: I agree. That is a better way to do that.

File: lib/resources/host.rb (hunk 1 above)
Comment (walker233) [reply lib/resources/host.rb:129]: That would probably be safe, but I thought for some reason it was nearly universal in inclusion. At least on redhat and debian families. Is that not the case. By the same token we are going to be safe with bash, we should probably do the same thing for the timeout command, which is in both redhat coreutils and debian coreutils packages. And were you suggesting we'd modify as:
~~~
if port && (inspect.os.redhat? || inspec.os.debian?) && 
   command('bash').exist? && command('timeout').exist?
~~~
Or were you thinking of a different modification.

**Comment by walker233** [comment]
@chris-rock I've been trying to get useful test cases but there is something I'm missing. I'm expecting a failure for 4 of the test cases when trying to connect to a tcp port on the host example.com, but I'm only getting positive indicators. I tried ports 8080 and now 9080 which all are returning true, when they are returning false in my environment. Any thoughts on why the host test come back as reachable?

**Comment by chris-rock** [comment]
@walker233 I tried the following quickly:
```
inspec> cmd = command("bash -c 'cat </dev/null >/dev/tcp/localhost/80'") 
=> Command bash -c 'cat </dev/null >/dev/tcp/localhost/80'
inspec> cmd.stdout
=> ""
inspec> cmd.stderr
=> "bash: connect: Connection refused\nbash: /dev/tcp/localhost/80: Connection refused\n"
inspec> cmd.exit_status
=> 1
```
I would expect to get an error back for any unassigned port. Isn't that working on your side?

**Comment by adamleff** [comment]
I haven't seen any movement on this PR, and we recently added support for using `nc` for testing TCP ports in the `host` resource in #1915. Therefore, I'm going to close this PR. Please let me know if we need to reopen this. Thank you!

**Pull request closed without merging**